File: app/src/main/java/com/example/asiantech/playmusic/fragment/BaseDetailFragment.java
package com.example.asiantech.playmusic.fragment;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.widget.TextView;

import com.example.asiantech.playmusic.MainActivity;
import com.example.asiantech.playmusic.MainActivity_;
import com.example.asiantech.playmusic.OnItemListener;
import com.example.asiantech.playmusic.R;
import com.example.asiantech.playmusic.adapter.SongAdapter;
import com.example.asiantech.playmusic.model.Song;
import com.example.asiantech.playmusic.service.MusicService;

import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.Click;
import org.androidannotations.annotations.EFragment;
import org.androidannotations.annotations.FragmentArg;
import org.androidannotations.annotations.ViewById;

import java.util.ArrayList;

/**
 * @author HoaHT
 */
@EFragment
public abstract class BaseDetailFragment extends Fragment implements OnItemListener {

    private static final String TAG = BaseDetailFragment.class.getSimpleName();

    @ViewById(R.id.recycleViewListSong)
    RecyclerView mRecycleListSong;

    @ViewById(R.id.tvTitle)
    TextView mTvTitle;

    @FragmentArg
    long albumId;

    protected MusicService mMusicService;
    private ArrayList<Song> mListSongAlbum;
    private SongAdapter mAdapter;

    @AfterViews
    void afterViews() {
        Log.d(TAG, "afterViews: ");
        ArrayList<Song> listSong = new ArrayList<>();
        if (getActivity() instanceof MainActivity_) {
            listSong = ((MainActivity_) getActivity()).getMListSong();
            mMusicService = ((MainActivity_) getActivity()).getMMusicService();
        }
        mListSongAlbum = new ArrayList<>();
        for (int i = listSong.size() - 1; i >= 0; i--) {
            Song song = listSong.get(i);
            if (getTypeId(song) == albumId) {
                mListSongAlbum.add(new Song(song.getId(), song.getTitle(), song.getArtistId(),
                        song.getArtist(), song.getAlbumId(), song.getAlbum(), song.getDisplay(),
                        false));
            }
        }
        mTvTitle.setText(getType(mListSongAlbum.get(0)));
        if (mMusicService.getListType().equals(MainActivity.LIST_TYPE_ALBUM)
                || mMusicService.getListType().equals(MainActivity.LIST_TYPE_ARTIST)) {
            for (int i = 0; i < mListSongAlbum.size(); i++) {
                if (mMusicService.getSongPlayingId() == mListSongAlbum.get(i).getId()) {
                    mListSongAlbum.get(i).setPlaying(true);
                }
            }
        }
        mAdapter = new SongAdapter(getContext(), mListSongAlbum, this);
        mRecycleListSong.setLayoutManager(new LinearLayoutManager(getContext()));
        mRecycleListSong.setAdapter(mAdapter);
        if (mReceiver != null) {
            IntentFilter intentFilter = new IntentFilter(MainActivity.ACTION_STRING_ACTIVITY);
            getActivity().registerReceiver(mReceiver, intentFilter);
        }
    }

    protected abstract long getTypeId(Song song);

    protected abstract String getType(Song song);

    protected abstract void setListType();

    private BroadcastReceiver mReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            Log.d(TAG, "onReceive: ");
            if (intent != null) {
                String message = intent.getExtras().getString("message");
                if (message != null) {
                    switch (message) {
                        case "play":
                            updateSongPlay();
                            break;
                    }
                }
            }
        }
    };

    private void updateSongPlay() {
        for (int i = 0; i < mListSongAlbum.size(); i++) {
            if (mListSongAlbum.get(i).getId() == mMusicService.getSongPlayingId()) {
                mListSongAlbum.get(i).setPlaying(true);
            } else {
                mListSongAlbum.get(i).setPlaying(false);
            }
        }
        mAdapter.notifyDataSetChanged();
    }

    @Click(R.id.imgGoBack)
    void onClickGoBack() {
        getActivity().onBackPressed();
    }

    @Override
    public void onDestroy() {
        if (mReceiver != null) {
            getActivity().unregisterReceiver(mReceiver);
        }
        super.onDestroy();
    }

    @Override
    public void onItemClick(int position) {
        Log.d(TAG, "onItemClick: ");
        mMusicService.setPlayList(mListSongAlbum);
        setListType();
        if (getActivity() instanceof MainActivity_) {
            ((MainActivity_) getActivity()).resetController();
        }
        mMusicService.setSong(position);
        mMusicService.playSong();
        updateSongPlay();
    }
}

File: app/src/main/java/com/example/asiantech/playmusic/fragment/BaseDetailFragment.kt
package com.example.asiantech.playmusic.fragment

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.example.asiantech.playmusic.MainActivity
import com.example.asiantech.playmusic.OnItemListener
import com.example.asiantech.playmusic.R
import com.example.asiantech.playmusic.adapter.SongAdapter
import com.example.asiantech.playmusic.model.Song
import com.example.asiantech.playmusic.service.MusicService
import kotlinx.android.synthetic.main.list_detail_fragment.*

/**
 * BaseDetailFragment.
 *
 * @author HoaHT
 */
abstract class BaseDetailFragment : Fragment(), OnItemListener {
    companion object {
        val KEY_ALBUM_ID = "key_album_id"
    }

    protected var albumId: Long = 0L
    protected var mMusicService: MusicService? = null
    private var listSongAlbum = ArrayList<Song>()
    private var adapter: SongAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        var listSong = ArrayList<Song>()
        albumId = arguments.getLong(KEY_ALBUM_ID)
        if (activity is MainActivity) {
            listSong = (activity as MainActivity).getListSong()
            mMusicService = (activity as MainActivity).getMusicService()
        }
        (listSong.size - 1 downTo 0)
                .map { listSong[it] }
                .filter { getTypeId(it) == albumId }
                .forEach {
                    listSongAlbum.add(Song(it.id, it.title, it.artistId, it.artist, it.albumId,
                            it.album, it.display, false))
                }
        if (mMusicService?.listType.equals(MainActivity.LIST_TYPE_ALBUM)
                || mMusicService?.listType.equals(MainActivity.LIST_TYPE_ARTIST)) {
            (0..listSongAlbum.size - 1)
                    .filter { mMusicService?.songPlayingId == listSongAlbum[it].id }
                    .forEach { listSongAlbum[it].isPlaying = true }
        }
        val intentFilter = IntentFilter(MainActivity.ACTION_STRING_ACTIVITY)
        activity.registerReceiver(receiver, intentFilter)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = LayoutInflater.from(context).inflate(R.layout.list_detail_fragment, container, false)
        return view
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        adapter = SongAdapter(context, listSongAlbum, this)
        recycleViewListSong.layoutManager = LinearLayoutManager(context)
        recycleViewListSong.adapter = adapter
        if (listSongAlbum.size > 0) {
            tvTitle.text = getType(listSongAlbum[0])
        }
        imgGoBack.setOnClickListener {
            activity.onBackPressed()
        }
    }

    protected abstract fun getTypeId(song: Song): Long

    protected abstract fun getType(song: Song): String

    protected abstract fun setListType()

    private val receiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            if (intent == null) {
                return
            }
            val message: String? = intent.extras.getString("message") ?: return
            when (message) {
                "play" -> updateSongPlay()
            }
        }
    }

    private fun updateSongPlay() {
        for (i in 0..listSongAlbum.size - 1) {
            listSongAlbum[i].isPlaying = listSongAlbum[i].id == mMusicService?.songPlayingId
        }
        adapter?.notifyDataSetChanged()
    }

    override fun onDestroy() {
        super.onDestroy()
        activity.unregisterReceiver(receiver)
    }

    override fun onItemClick(position: Int) {
        mMusicService?.playList = listSongAlbum
        setListType()
        if (activity is MainActivity) {
            (activity as MainActivity).resetController()
        }
        mMusicService?.setSong(position)
        mMusicService?.playSong()
        updateSongPlay()
    }
}